flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -957,17 +953,19 @@ private void rememberRecentCheckpointId(long id) {
 	}
 
 	private void dropSubsumedCheckpoints(long checkpointId) {
-		Iterator<Map.Entry<Long, PendingCheckpoint>> entries = pendingCheckpoints.entrySet().iterator();
-
-		while (entries.hasNext()) {
-			PendingCheckpoint p = entries.next().getValue();
-			// remove all pending checkpoints that are lesser than the current completed checkpoint
-			if (p.getCheckpointId() < checkpointId && p.canBeSubsumed()) {
-				rememberRecentCheckpointId(p.getCheckpointId());
-				failPendingCheckpoint(p, CheckpointFailureReason.CHECKPOINT_SUBSUMED);
-				entries.remove();
-			}
-		}
+		PendingCheckpoint[] checkpointsToSubsume =
+			pendingCheckpoints
+				.values()
+				.stream()
+				.filter(
+					pendingCheckpoint ->
+						pendingCheckpoint.getCheckpointId() < checkpointId &&
+						pendingCheckpoint.canBeSubsumed())
+				.toArray(PendingCheckpoint[]::new);
+
+		abortPendingCheckpoints(
+			checkpointsToSubsume,
+			new CheckpointException(CheckpointFailureReason.CHECKPOINT_SUBSUMED));
```

Comment:
For one thing, you could pass a stream as a parameter to the `abortPendingCheckpoints`, to deduplicate code a bit. However to be honest I do not see much value of keeping `abortPendingCheckpoints(PendingCheckpoint[] checkpoints, CheckpointException exception)` and I think I would just inline it.

Having it, deduplicates only a simple `for (PendingCheckpoint pendingCheckpoint : checkpoints) ` loop, while callers still have to iterate over the checkpoints one way or another (creating a stream, filtering it, collecting to an array).

Reply:
The purpose of passing an array to `abortPendingCheckpoints` as parameter is to avoid potential `ConcurrentModificationException`. There are more details in the response above.

Reply:
Do you mean providing a `abortPendingCheckpoint(Predicate<PendingCheckpint> checkpointToDeletePredicate)`,  instead of calling `pendingCheckpoints.remove(pendingCheckpoint.getCheckpointId())` directly in `abortPendingCheckpoint`?

I guess it works. However I'm not sure we should do that or not. The codes might be hard to understand. It seems that we have to make a decision about how to trade off this.

Reply:
I think I got your point now :)
Makes sense to me.